src/storage/src/table/state_table.rs

```diff
@@ -102,28 +111,118 @@ impl<S: StateStore> StateTable<S> {
         Ok(())
     }
 
-    pub async fn iter(&self, _pk: Row) -> StorageResult<StateTableRowIter<S>> {
-        todo!()
+    pub async fn iter(&self, epoch: u64) -> StorageResult<StateTableRowIter<'_, S>> {
+        let mem_table_iter = self.mem_table.buffer.iter().peekable();
+        StateTableRowIter::new(
+            &self.keyspace,
+            &self.column_descs,
+            mem_table_iter,
+            &self.order_types,
+            epoch,
+        )
+        .await
     }
 }
 
-pub struct StateTableRowIter<S: StateStore> {
-    cell_based_iter: CellBasedTableRowIter<S>,
-    mem_table: MemTable,
+/// `StateTableRowIter` is able to read the just written data (uncommited data).
+/// It will merge the result of `mem_table_iter` and `cell_based_streaming_iter`
+pub struct StateTableRowIter<'a, S: StateStore> {
+    mem_table_iter: Peekable<MemTableIter<'a>>,
+    cell_based_streaming_iter: CellBasedTableStreamingIter<S>,
+
+    /// The result of the last cell_based_streaming_iter next is saved, which can avoid being
+    /// discarded.
+    cell_based_item: Option<(Vec<u8>, Row)>,
+    pk_serializer: OrderedRowSerializer,
 }
+type MemTableIter<'a> = btree_map::Iter<'a, Row, RowOp>;
 
-impl<S: StateStore> StateTableRowIter<S> {
+impl<'a, S: StateStore> StateTableRowIter<'a, S> {
     async fn new(
-        _keyspace: Keyspace<S>,
-        _table_descs: Vec<ColumnDesc>,
-        _epoch: u64,
-        _stats: Arc<StateStoreMetrics>,
-    ) -> StorageResult<Self> {
-        todo!()
+        keyspace: &Keyspace<S>,
+        table_descs: &[ColumnDesc],
+        mem_table_iter: Peekable<MemTableIter<'a>>,
+        order_types_vec: &[OrderType],
+        epoch: u64,
+    ) -> StorageResult<StateTableRowIter<'a, S>> {
+        let cell_based_streaming_iter =
+            CellBasedTableStreamingIter::new(keyspace, table_descs, epoch).await?;
+        let pk_serializer = OrderedRowSerializer::new(order_types_vec.to_vec());
+        let state_table_iter = Self {
+            mem_table_iter,
+            cell_based_streaming_iter,
+            cell_based_item: None,
+            pk_serializer,
+        };
+
+        Ok(state_table_iter)
     }
 
-    async fn next(&mut self) -> StorageResult<Option<Row>> {
-        todo!()
+    pub async fn next(&mut self) -> StorageResult<Option<Row>> {
+        let mut mem_table_iter_next_flag = false;
+        let mut res = None;
+        if self.cell_based_item.is_none() {
+            self.cell_based_item = self.cell_based_streaming_iter.next().await.map_err(err)?;
+        }
+        match (self.cell_based_item.as_ref(), self.mem_table_iter.peek()) {
```

Comment:
If you remove the `as_ref`, will you still need to clone `self.cell_based_item`? Basically, I think you should not need to clone cell_based_item at all.

Reply:
See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d45f2bed60baaef4c61dac23085b28c0

Reply:
Indeed my fault! fixed.